check that version is defined before being considered a valid swagger…

redmachine2 opened the pull request:
… or openapi file.

I didn't add any new tests as I was not able to figure out how to mock the element to include the swagger key, if you have any suggestions that would be great.

fixes #161

redmachine2 committed:
check that version is defined before being considered a valid swagger or openapi file

semonte commented:
@redmachine2 Thanks for implementing this. It is not needed to write additional tests for this; there are hundreds of tests that load a Swagger spec so if the tests pass, the version check works as expected 👍

semonte commented:
I'll review the code soon.

semonte commented:
This looks great.

Could you still extend the test method `testNonSwaggerFileHasNoErrors()` in `ValidFileTest` so that the test Swagger spec has a key `swagger` without the version value. This would verify your initial issue.

semonte commented on src/main/java/org/zalando/intellij/swagger/file/FileDetector.java:
```diff
@@ -11,34 +12,46 @@
 
-        return JsonLanguage.INSTANCE.is(language) && new PathFinder().findByPathFrom(lookupKey, psiFile).isPresent();
+    private boolean hasVersion(final PsiElement psiElement, final String lookupVersion) {
+        return psiElement.getLastChild().getText().contains(lookupVersion);
```
We should use optionals here, `getLastChild()` and `getText()` can return null values.

semonte replied:
We could also check that the value is an exact match, according to spec the value must be `2.0` for Swagger. Open API has it similar.

redmachine2 committed:
update FileDetector hasVersion to use Optionals as getLastChild and getText can return null
verify that swagger version is "2.0"
verify that openapi version is using semantic versioning and starts with 3
update test files to also follow spec

semonte commented on src/main/java/org/zalando/intellij/swagger/file/FileDetector.java:
```diff
@@ -3,42 +3,60 @@
 
-        return JsonLanguage.INSTANCE.is(language) && new PathFinder().findByPathFrom(lookupKey, psiFile).isPresent();
+    private boolean hasVersion(final PsiElement psiElement, final String lookupVersion) {
+        return Optional.ofNullable(psiElement.getLastChild())
```
This looks overly complex for detecting the text, this would do:

```java
        return Optional.ofNullable(psiElement.getLastChild())
                .map(PsiElement::getText)
                .filter(text -> text.matches(lookupVersion))
                .isPresent();
```

redmachine2 replied:
Ok awesome. I've really just started using Streams and the first time I've used Optionals for anything. I'll update the logic to simplify it.

semonte replied:
Thanks, good to go 👍

semonte reviewed (commented):
Looks good, had one comment regarding the use of optionals.

redmachine2 committed:
simplify hasVersion logic to use a map
